Code/Framework/AzFramework/AzFramework/Terrain/TerrainDataRequestBus.h

```diff
@@ -26,6 +26,100 @@ namespace AzFramework
         typedef AZStd::function<void(const SurfaceData::SurfacePoint& surfacePoint, bool terrainExists)> SurfacePointListFillCallback;
 
+        //! Helper structure that defines a query region to use with the QueryRegion / QueryRegionAsync APIs.
+        struct TerrainQueryRegion
+        {
+            AZ_TYPE_INFO(TerrainQueryRegion, "{DE3F634D-9689-4FBC-9F43-A39CFDF425D0}");
+
+            TerrainQueryRegion() = default;
+            ~TerrainQueryRegion() = default;
+
+            TerrainQueryRegion(const AZ::Vector3& startPoint, size_t numPointsX, size_t numPointsY, const AZ::Vector2& stepSize);
+            TerrainQueryRegion(const AZ::Vector2& startPoint, size_t numPointsX, size_t numPointsY, const AZ::Vector2& stepSize);
+
+            static TerrainQueryRegion CreateFromAabbAndStepSize(const AZ::Aabb& region, const AZ::Vector2& stepSize);
+
+            AZ::Vector3 m_startPoint{ 0.0f }; //! The starting point for the region query
+            AZ::Vector2 m_stepSize{ 0.0f }; //! The step size to use for advancing to the next point to query
+            size_t m_numPointsX{ 0 }; //! The total number of points to query in the X direction
+            size_t m_numPointsY{ 0 }; //! The total number of points to query in the Y direction
+        };
+
+        //! A JobContext used to run jobs spawned by calls to the various Query*Async functions.
+        class TerrainJobContext : public AZ::JobContext
+        {
+        public:
+            TerrainJobContext(AZ::JobManager& jobManager, int numJobsToComplete)
+                : JobContext(jobManager)
+                , m_numJobsToComplete(numJobsToComplete)
+            {
+            }
+
+            // When a terrain job context is cancelled, all associated
+            // jobs are still guaranteed to at least begin processing,
+            // and if any QueryAsyncParams::m_completionCallback was
+            // set it's guaranteed to be called even in the event of a
+            // cancellation. If a job only begins processing after its
+            // associated job context has been cancelled, no processing
+            // will occur and the callback will be invoked immediately,
+            // otherwise the job may either run to completion or cease
+            // processing early; the callback is invoked in all cases,
+            // provided one was specified with the original request.
+            void Cancel()
+            {
+                m_isCancelled = true;
+            }
+
+            // Was this TerrainJobContext cancelled?
+            bool IsCancelled() const
+            {
+                return m_isCancelled;
+            }
+
+            // Called by the TerrainSystem when a job associated with
+            // this TerrainJobContext completes. Returns true if this
+            // was the final job to be completed, or false otherwise.
+            bool OnJobCompleted()
+            {
+                return (--m_numJobsToComplete == 0);
```

Comment:
Just for safety it'd be better to check `(--m_numJobsToComplete <= 0)` instead, since the variable is an integer.